operations/secure_random/secure_random.cc

```diff
@@ -0,0 +1,203 @@
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/kernels/bounds_check.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/lib/random/random_distributions.h"
+#include "sodium.h"
+
+using namespace tensorflow;
+
+using shape_inference::DimensionHandle;
+using shape_inference::ShapeHandle;
+
+#define CHACHABLOCKSIZE 64
+
+static Status SecureRandomShape(shape_inference::InferenceContext* context) {
+  // Check seed shape
+  ShapeHandle seed;
+  TF_RETURN_IF_ERROR(context->WithRank(context->input(1), 1, &seed));
+  DimensionHandle unused;
+  TF_RETURN_IF_ERROR(context->WithValue(context->Dim(seed, 0), 8, &unused));
+
+  // Set output shape
+  ShapeHandle out;
+  TF_RETURN_IF_ERROR(context->MakeShapeFromShapeTensor(0, &out));
+  context->set_output(0, out);
+  return Status::OK();
+}
+
+REGISTER_OP("SecureRandom")
+    .Input("shape: T")
+    .Input("seed: Tseed")
+    .Input("minval: dtype")
+    .Input("maxval: dtype")
+    .Output("output: dtype")
+    .Attr("dtype: {int32, int64} = DT_INT32")
+    .Attr("T: {int32, int64} = DT_INT32")
+    .Attr("Tseed: {int32} = DT_INT32")
+    .SetShapeFn(SecureRandomShape);
+
+// this function allows you to skip ahead in the chacha stream so you don't
+// have to allocate more memory than you need to, used in rejection sampling and
+// will be handy for parallelizing this operation
+void randombytes_buf_deterministic_ic(void * const buf, const size_t size, uint32_t count,
+                              const unsigned char seed[randombytes_SEEDBYTES])
+{
+    static const unsigned char nonce[crypto_stream_chacha20_ietf_NONCEBYTES] = {
+        'L', 'i', 'b', 's', 'o', 'd', 'i', 'u', 'm', 'D', 'R', 'G'
```

Comment:
that doesn't look like a nonce to me :P was it used like this elsewhere?

Reply:
Borrowed from libsodium. Here https://github.com/jedisct1/libsodium/blob/master/src/libsodium/randombytes/randombytes.c#L171...

Reply:
There's some information on the nonce here: https://download.libsodium.org/doc/advanced/stream_ciphers/chacha20

TL;DR: The stream cipher is created via the ietf version of the chacha20 stream cipher. Its based on a pair of (nonce, key) where the nonce is 96 bits and the key is 256 bits.